packages/backend/src/open_payments/grant/service.test.ts

```diff
@@ -201,4 +219,400 @@ describe('Grant Service', (): void => {
+        accessActions: options.accessActions,
+        authServerId: authServer.id,
+        accessToken: newOpenPaymentsGrant.access_token.value,
+        expiresAt: new Date(
+          Date.now() + newOpenPaymentsGrant.access_token.expires_in! * 1000
+        ),
+        managementId
+      })
+      expect(openPaymentsTokenRotationSpy).toHaveBeenCalled()
+      expect(openPaymentsGrantRequestSpy).toHaveBeenCalled()
+
+      const originalGrant = await Grant.query(knex).findById(existingGrant.id)
+      expect(originalGrant?.deletedAt).toBeDefined()
+    })
+
+    test('returns error if Open Payments grant request fails', async () => {
+      const openPaymentsGrantRequestSpy = jest
+        .spyOn(openPaymentsClient.grant, 'request')
+        .mockImplementationOnce(() => {
+          throw new Error('Could not request grant')
+        })
+
+      const options = {
+        authServer: authServer.url,
+        accessType: AccessType.IncomingPayment,
+        accessActions: [AccessAction.Create, AccessAction.ReadAll]
+      }
+
+      const error = await grantService.getOrCreate(options)
+
+      expect(error).toBe(GrantError.InvalidGrantRequest)
+      expect(openPaymentsGrantRequestSpy).toHaveBeenCalled()
+    })
+
+    test('returns error if Open Payments grant request returns a pending grant', async () => {
+      const openPaymentsGrantRequestSpy = jest
+        .spyOn(openPaymentsClient.grant, 'request')
+        .mockResolvedValueOnce(mockPendingGrant())
+
+      const options = {
+        authServer: authServer.url,
+        accessType: AccessType.IncomingPayment,
+        accessActions: [AccessAction.Create, AccessAction.ReadAll]
+      }
+
+      const error = await grantService.getOrCreate(options)
+
+      expect(error).toBe(GrantError.GrantRequiresInteraction)
+      expect(openPaymentsGrantRequestSpy).toHaveBeenCalled()
+    })
+  })
+
+  describe('getExistingGrant', (): void => {
+    let authServer: AuthServer
+
+    beforeEach(async (): Promise<void> => {
+      const authServerService = await deps.use('authServerService')
+      const url = faker.internet.url({ appendSlash: false })
+      authServer = await authServerService.getOrCreate(url)
+    })
+
+    test('gets existing grant (identical match)', async () => {
+      const existingGrant = await Grant.query(knex).insertAndFetch({
+        authServerId: authServer.id,
+        accessToken: uuid(),
+        managementId: uuid(),
+        accessType: AccessType.IncomingPayment,
+        accessActions: [AccessAction.ReadAll]
+      })
+
+      const options = {
+        authServer: authServer.url,
+        accessType: AccessType.IncomingPayment,
+        accessActions: [AccessAction.ReadAll]
+      }
+
+      await expect(
+        getExistingGrant({ knex } as ServiceDependencies, options)
+      ).resolves.toEqual({ ...existingGrant, authServer })
+    })
+
+    test('gets existing grant (requested actions are a subset of saved actions)', async () => {
```

Comment:
Would be necessary to create another test (or modify this one) to ensure that it can retrieve existing grants that have (Read/List)-All rights when (Read/List) is requested?

Reply:
good point, maybe during create if ReadAll/ListAll is requested, we can also manually add the corresponding Read/List as well